crates/block-producer/src/block_builder/prover/asm/block_kernel.masm

```diff
@@ -13,154 +13,228 @@ const.CHAIN_MMR_PTR=1000
+        # OS => [OLD_VALUE, ROOT_{i+1}, NULLIFIER_VALUE, counter]
+        # AS => [[NULLIFIER_{i+1}]]
 
         # Check that OLD_VALUE == 0 (i.e. that nullifier was indeed not previously produced)
         assertz assertz assertz assertz
-        #=> [ROOT_{i+1}, NULLIFIER_VALUE, num_nullifiers_left_to_update, ... ]
+        # OS => [ROOT_{i+1}, NULLIFIER_VALUE, counter]
+        # AS => [[NULLIFIER_{i+1}]]
 
         # loop counter
-        movup.8 sub.1 dup neq.0
-        #=> [0 or 1, num_nullifiers_left_to_update - 1, ROOT_{i+1}, NULLIFIER_VALUE, ... ]
+        movup.8 sub.1 dup movdn.9 neq.0
+        # OS => [flag, ROOT_{i+1}, NULLIFIER_VALUE, counter]
+        # AS => [[NULLIFIER_{i+1}]]
     end
-    #=> [0, ROOT_{n-1}, NULLIFIER_VALUE ]
 
-    drop swapw dropw
-    # => [ROOT_{n-1}]
+    # drop the counter and the nullifier value
+    swapw dropw movup.4 drop
+    # OS => [ROOT_{n-1}]
+    # AS => []
 end
 
 #! Compute the chain MMR root
 #!
-#! Stack: [ PREV_CHAIN_MMR_HASH, PREV_BLOCK_HASH_TO_INSERT ]
-#! Advice map: PREV_CHAIN_MMR_HASH -> NUM_LEAVES || peak_0 || .. || peak_{n-1} || <maybe padding>
-#!
-#! Output: [ CHAIN_MMR_ROOT ]
+#! Inputs:
+#!   Operand stack: []
+#!   Advice stack:  [PREV_BLOCK_HASH_TO_INSERT, PREV_CHAIN_MMR_HASH]
+#!   Advice map: {
+#!      PREV_CHAIN_MMR_HASH: [NUM_LEAVES, [peak_i], <maybe padding>]
+#!   }
+#! Outputs:
+#!   Operand stack: [CHAIN_MMR_ROOT]
 proc.compute_chain_mmr_root
+    # move the previous block hash and chain MMR hash to the operand stack
+    adv_push.8
+    # OS => [PREV_CHAIN_MMR_HASH, PREV_BLOCK_HASH_TO_INSERT]
+    # AS => []
+
+    # push chain MMR pointer to the operand stack
     push.CHAIN_MMR_PTR movdn.4
-    # => [ PREV_CHAIN_MMR_HASH, chain_mmr_ptr, PREV_BLOCK_HASH_TO_INSERT ]
+    # OS => [PREV_CHAIN_MMR_HASH, chain_mmr_ptr, PREV_BLOCK_HASH_TO_INSERT]
 
     # load the chain MMR (as of previous block) at memory location CHAIN_MMR_PTR
     exec.mmr::unpack
-    # => [ PREV_BLOCK_HASH_TO_INSERT ]
+    # OS => [PREV_BLOCK_HASH_TO_INSERT]
 
+    # push chain MMR pointer to the operand stack
     push.CHAIN_MMR_PTR movdn.4
-    # => [ PREV_BLOCK_HASH_TO_INSERT, chain_mmr_ptr ]
+    # OS => [PREV_BLOCK_HASH_TO_INSERT, chain_mmr_ptr]
 
     # add PREV_BLOCK_HASH_TO_INSERT to chain MMR
     exec.mmr::add
-    # => [ ]
+    # OS => []
 
     # Compute new MMR root
     push.CHAIN_MMR_PTR exec.mmr::pack
-    # => [ CHAIN_MMR_ROOT ]
+    # OS => [CHAIN_MMR_ROOT]
 end
 
-# Stack: [<account root inputs>, <note root inputs>, <nullifier root inputs>, <chain mmr root inputs>]
+#! Inputs:
+#!   Operand stack: []
+#!   Advice stack:  [<account root inputs>, <note root inputs>, <nullifier root inputs>, <chain mmr root inputs>]
+#!   Advice map: {
+#!     PREV_CHAIN_MMR_HASH: [NUM_LEAVES, [peak_i], <maybe padding>]
+#!   }
+#! Outputs:
+#!   Operand stack: [ACCOUNT_ROOT, NOTE_ROOT, NULLIFIER_ROOT, CHAIN_MMR_ROOT]
 begin
```

Comment:
This is fine for now, but we should be providing some inputs via the operand stack - otherwise this can be used to provide any arbitrary batch inputs.

Let's create 2 issues in `miden-base` for this: one issue for defining batch kernel inputs/outputs, and the other one for defining block kernel inputs/outputs.

Reply:
@bobbinth I created two issue templates: https://github.com/0xPolygonMiden/miden-base/issues/919 and https://github.com/0xPolygonMiden/miden-base/issues/920, although didn't manage to fill them properly, I'm not sure I understand what should be done.